internal/storage/fs/git/store.go

```diff
@@ -256,6 +259,40 @@ func (s *SnapshotStore) resolve(ref string) (plumbing.Hash, error) {
 		return plumbing.NewHash(ref), nil
 	}
 
+	constraint, err := semver.NewConstraint(ref)
+	if err == nil { // If there are no errors it's a valid semver constraint
```

Comment:
There might be a problem with the `v1` or similar branch as a reference. Since this constraint prioritizes tags, this could lead using the wrong configuration or errors altogether.

Reply:
yeah looping in @GeorgeMac here as well as he initially implemented this

Reply:
Actually `v1` is not a valid semver version/constraint so it won't be a problem! You can check it here [constraintRegex](https://github.com/Masterminds/semver/blob/master/constraints.go#L186)

The problem will be if there's a branch named `v1.0.0` or `1.0.0`

Reply:
```go
package main

import (
	"fmt"

	"github.com/Masterminds/semver/v3"
)

func main() {
	_, err := semver.NewConstraint("v1-api")
	if err != nil {
		fmt.Printf("failed to parse version 1: %v\n", err)
		return
	}
	fmt.Println("valid")
}
```
Running it...
```sh
❯ go run main.go
valid
```

Reply:
IIRC there is a parse strict version we could use?

Reply:
Flipt uses `github.com/blang/semver/v4` and `golang.org/x/mod/semver` currently. `github.com/Masterminds/semver/v3` is a third dep. Each one has own flavors. Masterminds is fine with `v` as prefix, blang is more strict about this. Using any dep will require some extra complexity.

Just to remember that `1.2.3-beta.1+build345` is a valid semver, this is less than `1.2.3` but higher than `1.2.2`. The documentation should be very good for this feature to remove misuse.

Reply:
I haven't seen Flipt was already using other libraries! I think we can stick with `github.com/blang/semver/v4` because `golang.org/x/mod/semver` is also not strict

This is from mod/semver documentation:
> The general form of a semantic version string accepted by this package is
> ```text
> vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]]
> ```

Reply:
I had a little walk around some projects. FluxCD uses this mastermind one for the same purpose:
https://github.com/fluxcd/flux2/blob/5456635ee7e13c551ec4e11242cd9beeceb838be/cmd/flux/check.go#L179

I can understand the need for this particular library here, as we want semver _constraints_, which supports a whole host of wildcard and range syntaxes, not valid semver itself. So while it is adding another semver dependency, I think this makes sense here.

https://github.com/Masterminds/semver?tab=readme-ov-file#checking-version-constraints

That said though, you're right that this could lead to some potentially unexpected behaviour. I am unsure what might end up constituting this valid semver constraint language.

What if we made this an alternative configuration property to `ref`, i.e. something like `semver` on the git configuration? Then make these mutually exclusive ways to configure this storage backend?

Reply:
Yeah, I've used it because I found it while contributing to ArgoCD!

I like your idea @GeorgeMac, but maybe we can have a "ref_type"?
```yaml
storage:
  type: git
  git:
    repository: "https://github.com/predictab.le/config.git"
    ref: "main"
    // "static" as a default value, "semver" indicates the user wants to resolve a semantic versioning constraint
    ref_type: "static"
```

Reply:
That's a great idea! saves doing fiddly mutually exclusive validation! Nice one 💪

Reply:
Great then, I'm gonna try to implement it by this weekend!

Reply:
no rush, but if you do manage to get it in this weekend we'll likely create a release on Monday 🚀

Reply:
I've pushed the changes!